subprojects/messaging/src/main/java/org/gradle/internal/event/DefaultListenerManager.java

```diff
@@ -153,63 +166,108 @@ public void dispatch(MethodInvocation message) {
             }
             return source.getSource();
         }
 
-        // Must be holding lock
         private void invalidateDispatchCache() {
             ensureAllWithLoggerInitialized();
             ensureAllWithoutLoggerInitialized();
         }
 
-        // Must be holding lock
-        void maybeAdd(ListenerDetails listener) {
+        void maybeAdd(final ListenerDetails listener) {
             if (type.isInstance(listener.listener)) {
-                listeners.add(listener);
-                invalidateDispatchCache();
+                if (broadcasterLock.tryLock()) {
+                    try {
+                        listeners.add(listener);
+                        invalidateDispatchCache();
+                    } finally {
+                        broadcasterLock.unlock();
+                    }
+                } else {
+                    synchronized (queuedOperations) {
+                        queuedOperations.add(new Runnable() {
+                            @Override
+                            public void run() {
+                                listeners.add(listener);
+                            }
+                        });
+                    }
+                }
             }
         }
 
-        // Must be holding lock
-        void maybeRemove(ListenerDetails listener) {
-            if (listeners.remove(listener)) {
-                invalidateDispatchCache();
+        void maybeRemove(final ListenerDetails listener) {
+            if (broadcasterLock.tryLock()) {
+                try {
+                    if (listeners.remove(listener)) {
+                        invalidateDispatchCache();
+                    }
+                } finally {
+                    broadcasterLock.unlock();
+                }
+            } else {
+                synchronized (queuedOperations) {
+                    queuedOperations.add(new Runnable() {
+                        @Override
+                        public void run() {
+                            listeners.remove(listener);
+                        }
+                    });
+                }
             }
-            // Another thread may be using listener
         }
 
-        // Must be holding lock
-        void maybeSetLogger(ListenerDetails candidate) {
+        void maybeSetLogger(final ListenerDetails candidate) {
             if (type.isInstance(candidate.listener)) {
-                if (logger == null && parent != null) {
-                    parentDispatch = parent.getBroadcasterInternal(type).getDispatch(false);
+                if (broadcasterLock.tryLock()) {
+                    try {
+                        doSetLogger(candidate);
+                        invalidateDispatchCache();
+                    } finally {
+                        broadcasterLock.unlock();
+                    }
+                } else {
+                    synchronized (queuedOperations) {
+                        queuedOperations.add(new Runnable() {
```

Comment:
I don't think so, we broadcast one message at a time for multiple listeners, and the lock is acquired for each message, so the queue would be cleaned after each message, if any message queued (it should be very rare).